src/particles/PhysicsParticles.hpp

```diff
@@ -0,0 +1,206 @@
+template <typename problem_t> constexpr int RadParticleRealComps = 3 + Physics_Traits<problem_t>::nGroups;
+template <typename problem_t> using RadParticleContainer = amrex::AmrParticleContainer<RadParticleRealComps<problem_t>>;
+template <typename problem_t> using RadParticleIterator = amrex::ParIter<RadParticleRealComps<problem_t>>;
+
+enum CICParticleDataIdx { CICParticleMassIdx = 0, CICParticleVxIdx, CICParticleVyIdx, CICParticleVzIdx };
+constexpr int CICParticleRealComps = 4; // mass vx vy vz
+using CICParticleContainer = amrex::AmrParticleContainer<CICParticleRealComps>;
+using CICParticleIterator = amrex::ParIter<CICParticleRealComps>;
+
+struct MassDeposition {
+	amrex::Real Gconst{};
+	int start_part_comp{};
+	int start_mesh_comp{};
+	int num_comp{};
+
+	template <typename ParticleType>
+	AMREX_GPU_DEVICE AMREX_FORCE_INLINE void operator()(const ParticleType &p, amrex::Array4<amrex::Real> const &rho,
+							    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> const &plo,
+							    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> const &dxi) const noexcept
+	{
+		amrex::ParticleInterpolator::Linear interp(p, plo, dxi);
+		interp.ParticleToMesh(p, rho, start_part_comp, start_mesh_comp, num_comp, [=] AMREX_GPU_DEVICE(const ParticleType &part, int comp) {
+			return 4.0 * M_PI * Gconst * part.rdata(comp); // weight by 4 pi G
+		});
+	}
+};
+
+struct RadDeposition {
+	double current_time{};
+	int start_part_comp{};
+	int start_mesh_comp{};
+	int num_comp{};
+
+	template <typename ParticleType>
+	AMREX_GPU_DEVICE AMREX_FORCE_INLINE void operator()(const ParticleType &p, amrex::Array4<amrex::Real> const &radEnergySource,
+							    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> const &plo,
+							    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> const &dxi) const noexcept
+	{
+		amrex::ParticleInterpolator::Linear interp(p, plo, dxi);
+		interp.ParticleToMesh(p, radEnergySource, start_part_comp, start_mesh_comp, num_comp, [=] AMREX_GPU_DEVICE(const ParticleType &part, int comp) {
+			if (current_time < part.rdata(RadParticleBirthTimeIdx) || current_time >= part.rdata(RadParticleDeathTimeIdx)) {
+				return 0.0;
+			}
+			return part.rdata(comp) * (AMREX_D_TERM(dxi[0], *dxi[1], *dxi[2]));
+		});
+	}
+};
+
+// Forward declarations
+template <typename problem_t> class PhysicsParticleRegister;
+
+// Base class for physics particle descriptors
+class PhysicsParticleDescriptor
+{
+      protected:
+	int massIndex_{-1};		 // index for gravity mass, -1 if not used
+	int lumIndex_{-1};		 // index for radiation luminosity, -1 if not used
+	bool interactsWithHydro_{false}; // whether particles interact with hydro
+
+      public:
+	PhysicsParticleDescriptor(int mass_idx, int lum_idx, bool hydro_interact)
+	    : massIndex_(mass_idx), lumIndex_(lum_idx), interactsWithHydro_(hydro_interact)
+	{
+	}
+	virtual ~PhysicsParticleDescriptor() = default;
+	void *neighborParticleContainer_{}; // pointer to particle container, type-erased
```

Comment:
There is no reason to do type-erasure here, and no reason to make this a raw pointer.

Reply:
You can make this `std::unique_ptr<amrex::ParticleContainerBase>` and then `dynamic_cast` to the sub-types of particle container, as needed: https://amrex-codes.github.io/amrex/doxygen/classamrex_1_1ParticleContainerBase.html